server/config.md
# Edumeet Server Config

| Property | Description | Format | Default value |
| :------- | :---------- | :----- | :------------ |
| turnAPIKey | TURN server key for requesting a geoip-based TURN server closest to the client. | `"string"` | ``""`` |
| turnAPIURI | TURN server URL for requesting a geoip-based TURN server closest to the client. | `"url"` | ``""`` |
| turnAPIparams.uri_schema | TURN server URL schema. | `"string"` | ``"turn"`` |
| turnAPIparams.transport | TURN server transport. | `[  "tcp",  "udp"]` | ``"tcp"`` |
| turnAPIparams.ip_ver | TURN server IP version. | `[  "ipv4",  "ipv6"]` | ``"ipv4"`` |
| turnAPIparams.servercount | TURN server count. | `"nat"` | ``2`` |
| turnAPITimeout | TURN server API timeout (seconds). | `"nat"` | ``2000`` |
| backupTurnServers | Backup TURN servers if REST fails or is not configured | `"*"` | ``[  {    "urls": [      "turn:turn.example.com:443?transport=tcp"    ],    "username": "example",    "credential": "example"  }]`` |
| fileTracker | Bittorrent tracker. | `"string"` | ``"wss://tracker.lab.vvc.niif.hu:443"`` |
| redisOptions | Redis server options. | `"object"` | ``{}`` |
| cookieSecret | Session cookie secret. | `"string"` | ``"T0P-S3cR3t_cook!e"`` |
| cookieName | Session cookie name. | `"string"` | ``"edumeet.sid"`` |
| tls.cert | SSL certificate path. | `"string"` | ``"./certs/mediasoup-demo.localhost.cert.pem"`` |
| tls.key | SSL key path. | `"string"` | ``"./certs/mediasoup-demo.localhost.key.pem"`` |
| listeningHost | The listening Host or IP address. If omitted listens on every IP. ("0.0.0.0" and "::"). | `"string"` | ``null`` |
| listeningPort | The HTTPS listening port. | `"port"` | ``8443`` |
| listeningRedirectPort | The HTTP listening port. Any HTTP request is redirected to HTTPS. | `"port"` | ``8080`` |
| httpOnly | Listens only on HTTP on listeningPort; listeningRedirectPort disabled. Use case: load balancer backend. | `"boolean"` | ``false`` |
| trustProxy | WebServer/Express trust proxy config for httpOnly mode. More infos: https://expressjs.com/en/guide/behind-proxies.html, https://www.npmjs.com/package/proxy-addr | `"string"` | ``""`` |
| activateOnHostJoin | When true, the room will be open to all users since there are users in the room. | `"boolean"` | ``true`` |
| roomsUnlocked | An array of rooms users can enter without waiting in the lobby. | `"array"` | ``null`` |
| maxUsersPerRoom | It defines how many users can join a single room. If not set, no limit is applied. | `"nat"` | ``null`` |
| routerScaleSize | Room size before spreading to a new router. | `"nat"` | ``40`` |
| requestTimeout | Socket timeout value (ms). | `"nat"` | ``20000`` |
| requestRetries | Socket retries when a timeout occurs. | `"nat"` | ``3`` |
| mediasoup.numWorkers | The number of Mediasoup workers to spawn. | `"nat"` | ``8`` |
| mediasoup.worker.logLevel | The Mediasoup log level. | `"string"` | ``"warn"`` |
| mediasoup.worker.logTags | The Mediasoup log tags. | `"array"` | ``[  "info",  "ice",  "dtls",  "rtp",  "srtp",  "rtcp"]`` |
| mediasoup.worker.rtcMinPort | The Mediasoup start listening port number. | `"port"` | ``40000`` |
| mediasoup.worker.rtcMaxPort | The Mediasoup end listening port number. | `"port"` | ``49999`` |
| mediasoup.router.mediaCodecs | The Mediasoup codecs settings. | `"object"` | ``[  {    "kind": "audio",    "mimeType": "audio/opus",    "clockRate": 48000,    "channels": 2  },  {    "kind": "video",    "mimeType": "video/VP8",    "clockRate": 90000,    "parameters": {      "x-google-start-bitrate": 1000    }  },  {    "kind": "video",    "mimeType": "video/VP9",    "clockRate": 90000,    "parameters": {      "profile-id": 2,      "x-google-start-bitrate": 1000    }  },  {    "kind": "video",    "mimeType": "video/h264",    "clockRate": 90000,    "parameters": {      "packetization-mode": 1,      "profile-level-id": "4d0032",      "level-asymmetry-allowed": 1,      "x-google-start-bitrate": 1000    }  },  {    "kind": "video",    "mimeType": "video/h264",    "clockRate": 90000,    "parameters": {      "packetization-mode": 1,      "profile-level-id": "42e01f",      "level-asymmetry-allowed": 1,      "x-google-start-bitrate": 1000    }  }]`` |
| mediasoup.webRtcTransport.listenIps | The Mediasoup listen IPs. https://mediasoup.org/documentation/v3/mediasoup/api/#TransportListenIp | `"array"` | ``[  {    "ip": "0.0.0.0",    "announcedIp": null  }]`` |
| mediasoup.webRtcTransport.initialAvailableOutgoingBitrate | The Mediasoup initial available outgoing bitrate (in bps). https://mediasoup.org/documentation/v3/mediasoup/api/#WebRtcTransportOptions | `"nat"` | ``1000000`` |
| mediasoup.webRtcTransport.maxIncomingBitrate | The Mediasoup maximum incoming bitrate for each transport. (in bps). https://mediasoup.org/documentation/v3/mediasoup/api/#transport-setMaxIncomingBitrate | `"nat"` | ``1500000`` |
| prometheus.deidentify | De-identify IP addresses in Prometheus logs. | `"boolean"` | ``false`` |
| prometheus.listen | Prometheus exporter listening address. | `"string"` | ``"localhost"`` |
| prometheus.numeric | Show numeric IP addresses in Prometheus logs. | `"boolean"` | ``false`` |
| prometheus.port | The Prometheus exporter listening port. | `"port"` | ``8889`` |
| prometheus.quiet | Include fewer labels in Prometheus logs. | `"boolean"` | ``false`` |
| prometheus.period | The Prometheus exporter update period (seconds). | `"nat"` | ``15`` |
| prometheus.secret | The Prometheus exporter authorization header: `Bearer <secret>` required to allow scraping. | `"string"` | ``null`` |


---

*Document generated with:* `yarn gen-config-docs`


CHANGELOG.md
# Changelog

## 3.5.0 or Next Version

### Added

* Add QoS set DSCP bits in IP header according to new option networkPriorities in app config
  Our media traffic should get now higher priority where QoS has been implemented.
  Routers will add higher priority or put media packets to low latency queue.
  If QoS implemented in L2 too, then switches also set the L2 CoS bits according DSCP.
  WIFI AP-s that implements QoS, also based on it could activate the multimedia extension MMS.
  Hopefully it will help and add little bit more priority lower latency in congested networks.
  At the time of this writing it is only implemented in Chrome.

## 3.4.0

### Added

* Multiparty meeting renamed to edumeet
* Merged room selector- and join- dialog - default keyboard focus on login field
* Localization selectable
* Added aspect ratio 16 : 9  and this is default now, ref #439
* New worker load calculation so router selection is based on that and not random anymore
* New permissions and roles:
  * Propagate userRoles to client state, ref #437
  * Extend userRoles and use the new audio, video permissions, ref #437
  * New permission to modify peer roles
  * Create room actions for giving and taking roles
  * Add functions to client for modifying roles live
  * Ability to give roles to users
  * Add new permission to config
* Add room to userMapping. Example of giving moderator if there is no authenticated user
* Promote all peers from lobby when a peer joins with the PROMOTE_PEER permission and activateOnHostJoin is true in config
* Make menus more intuitive on mobile
* Simplify electron screenshare check
* Logo support
* Improve autoMute mic-indicators, Improve audio level scaling
* Clean up participant list
* Add indicator for peers in focus. Ref #360
* TCP enabled by default, prefer UDP
* Ability for Prometheus exporter to listen on localhost
* Make list headers bolder
* Trim displayName inputs. Add random number to Guest displayName.
* Removed facingMode from mobile
* Documentation for prometheus exporter
* Added switching of own video mirror in settings
* Added hiding of own videos
* Request audio and webcam permission at once, when user is requesting media
* Add initial support for local and saml auth
  * Local login form
  * Add bcrypt encrypted passwords for local strategy
  * Add displayName mapping to usermapping
  * Add saml attriute mappings
* Use shared cookieparser for web and websocket
* Update TopBar leave button
* Add joinAudio capability
* Standardize Auth button

### Migration

from last master version:
* Copy paste defaultAudio and centralAudioOtions from config.example.js to config.js 
* Copy paste whole theme from config.example.js to config
* Configure logo in config.js

### Upgrade depencies

* webtorrent from 0.107.17 to 0.108.1 …
* Upgrade React-scripts

### Languages

* Updates translations: hu, tr, no, pl, uk
* Addad translations: Hindi (hi), Russian (ru), Kazakh (ka)

### Bugfixes

* Hopefully fix to silent peer issue, ref #256
* Set timeout for TURN API request, fixes #484
* Possible fix for #582 Crackling sound
* Fix for #444 Settings persistence
* Fix Audio settings from config.js take no effect
* Fix only firefox handles applyConstraints to audio tracks correctly
* Fix (autoMuted) mic too big click area
* Fix userRoles check
* Fix spoltights ignoring maxLastN
* Fix wrong config parameter naming (voiceActivatedUnmute)
* Fix locale checking state
* Fix express error handler
* Add constraint for user facing camera. Update media devices properly
* Cleanup on close
* Fixed null values for peers not yet assigned to a router
* Fix URL sanitizer bug
* Removed audio request modification
* Fix race in spotlights
* Fix piping bug when peers returning to router
* Fix piped router count
* Add comma to prometheus config code
* Removed code that is problematic and also unused
* Remove duplicated callbackURL
* Fix moderator buttons layout
* Fix: Special chars are not sanitized in URL
* Fix another roomId bug
* Tidy: replace obj.entries with obj.values to avoid unused key
* Fix close room link

## 3.3

* Add: Rooms now scale across cores
* Add: Permissions and roles. Users can now have different roles (moderator, admin etc.) that give different permissions.
* Add: TURN API or fallback TURN server
* Add: Configurable room size limit
* Add: Prometheus monitoring support
* Add: Possible to share several videos (ex: 2 webcams)
* Add: Configurable audio settings (echocancellation etc.)
* Add: Configurable audio output device (in supported browsers)
* Add: Audio auto mute/unmute based on volume
* Add: Handle unsupported browsers properly
* Add: Lots of appearance settings
* Add: Side drawer can now stay permanently open
* Add: Move control buttons to separate control bar
* Add: Can now "raise hand"
* Add: Screen sharing in Safari 13+, Opera and Edge
* Add: Extended advanced info about network in client
* Add: Configurable screen sharing frame rate
* Add: Help and About dialogs
* Add: More keyboard shortcuts
* Add: Quality indicator on videos
* Add: More translations
* Fix: Various UI fixes and improvements
* Fix: Better audio/video device handling
* Fix: Update keyboard shortcut handling
* Fix: Authentication for load balanced scenarios
* Fix: Signaling when entering lobby
* Fix: Signaling timeouts and retries
* Fix: Filesharing fixes (sharing same file twice, etc.)
* Fix: Better handling of hark
* Fix: Use applyContraints instead of restarting producers
* Fix: Now handles reconnects properly if client loses connection
* Fix: Rotating devices don't show rotated videos
* Fix: Various fixes to client authentication

## 3.2.1

* Fix: permananent top bar by default
* Fix: `httpOnly` mode https redirect
* Add some extra checks for video stream and track
* Add Italian translation
* Add Czech translation
* Add new server option `trustProxy` for load balancing http only use case
* Add HAproxy load balance example
* Add LTI LMS integration documentation
* Fix spacing of leave button

server/lib/config.ts
import * as fs from 'fs';
import convict from 'convict';
import { ipaddress, url } from 'convict-format-with-validator';
import json5 from 'json5';
import yaml from 'yaml';
import toml from 'toml';
import { cpus } from 'os';

import Logger from './Logger';

import * as userRoles from '../userRoles';
import {
	BYPASS_ROOM_LOCK,
	BYPASS_LOBBY
} from '../access';
import {
	CHANGE_ROOM_LOCK,
	PROMOTE_PEER,
	MODIFY_ROLE,
	SEND_CHAT,
	MODERATE_CHAT,
	SHARE_AUDIO,
	SHARE_VIDEO,
	SHARE_SCREEN,
	EXTRA_VIDEO,
	SHARE_FILE,
	MODERATE_FILES,
	MODERATE_ROOM
} from '../permissions';

const logger = new Logger('config');

// add parsers
convict.addParser([
	{ extension: 'json', parse: JSON.parse },
	{ extension: 'json5', parse: json5.parse },
	{ extension: [ 'yml', 'yaml' ], parse: yaml.parse },
	{ extension: 'toml', parse: toml.parse }
]);

// add formats
function assert(assertion: Boolean, msg: string)
{
	if (!assertion)
		throw new Error(msg);
}

const isFloat = {
	name     : 'float',
	coerce   : (v: string) => parseFloat(v),
	validate : (v: number) => assert(Number.isFinite(v), 'must be a number')
};

convict.addFormats({ ipaddress, url, isFloat });

// config schema
const configSchema = convict({
	turnAPIKey :
	{
		doc     : 'TURN server key for requesting a geoip-based TURN server closest to the client.',
		format  : String,
		default : ''
	},
	turnAPIURI :
	{
		doc     : 'TURN server URL for requesting a geoip-based TURN server closest to the client.',
		format  : String,
		default : ''
	},
	turnAPIparams :
	{
		'uri_schema' : {
			doc     : 'TURN server URL schema.',
			format  : String,
			default : 'turn'
		},
		'transport' : {
			doc     : 'TURN server transport.',
			format  : [ 'tcp', 'udp' ],
			default : 'tcp'
		},
		'ip_ver' : {
			doc     : 'TURN server IP version.',
			format  : [ 'ipv4', 'ipv6' ],
			default : 'ipv4'
		},
		'servercount' : {
			doc     : 'TURN server count.',
			format  : 'nat',
			default : 2
		}
	},
	turnAPITimeout : {
		doc     : 'TURN server API timeout (seconds).',
		format  : 'nat',
		default : 2 * 1000
	},
	backupTurnServers : {
		doc     : 'Backup TURN servers if REST fails or is not configured',
		format  : '*',
		default : [
			{
				urls : [
					'turn:turn.example.com:443?transport=tcp'
				],
				username   : 'example',
				credential : 'example'
			}
		]
	},
	fileTracker : {
		doc     : 'Bittorrent tracker.',
		format  : String,
		default : 'wss://tracker.lab.vvc.niif.hu:443'
	},
	redisOptions : {
		host : {
			doc     : 'Redis server host.',
			format  : String,
			default : 'localhost'
		},
		port : {
			doc     : 'Redis server port.',
			format  : 'port',
			default : 6379
		},
		password : {
			doc     : 'Redis server password.',
			format  : String,
			default : ''
		}
	},
	cookieSecret : {
		doc     : 'Session cookie secret.',
		format  : String,
		default : 'T0P-S3cR3t_cook!e'
	},
	cookieName : {
		doc     : 'Session cookie name.',
		format  : String,
		default : 'edumeet.sid'
	},
	tls : {
		cert : {
			doc     : 'SSL certificate path.',
			format  : String,
			default : './certs/mediasoup-demo.localhost.cert.pem'
		},
		key : {
			doc     : 'SSL key path.',
			format  : String,
			default : './certs/mediasoup-demo.localhost.key.pem'
		}
	},
	listeningHost : {
		doc     : 'The listening Host or IP address. If omitted listens on every IP. ("0.0.0.0" and "::").',
		format  : String,
		default : ''
	},
	listeningPort : {
		doc     : 'The HTTPS listening port.',
		format  : 'port',
		default : 8443
	},
	listeningRedirectPort : {
		doc     : 'The HTTP server listening port used for redirecting any HTTP request to HTTPS. If 0, the redirect server is disabled.',
		format  : 'port',
		default : 8080
	},
	httpOnly : {
		doc     : 'Listens only on HTTP on listeningPort; listeningRedirectPort disabled. Use case: load balancer backend.',
		format  : 'Boolean',
		default : false
	},
	trustProxy : {
		doc     : 'WebServer/Express trust proxy config for httpOnly mode. More infos: [expressjs](https://expressjs.com/en/guide/behind-proxies.html), [proxy-addr](https://www.npmjs.com/package/proxy-addr)',
		format  : String,
		default : ''
	},
	staticFilesCachePeriod : {
		doc     : 'The max-age in milliseconds for HTTP caching of static resources. This can also be a string accepted by the [ms module](https://www.npmjs.com/package/ms#readme).',
		format  : '*',
		default : 0
	},
	activateOnHostJoin : {
		doc     : 'When true, the room will be open to all users since there are users in the room.',
		format  : 'Boolean',
		default : true
	},
	roomsUnlocked : {
		doc     : 'An array of rooms users can enter without waiting in the lobby.',
		format  : Array,
		default : []
	},
	maxUsersPerRoom : {
		doc     : 'It defines how many users can join a single room. If not set, no limit is applied.',
		format  : 'nat',
		default : 0
	},
	routerScaleSize : {
		doc     : 'Room size before spreading to a new router.',
		format  : 'nat',
		default : 40
	},
	requestTimeout : {
		doc     : 'Socket timeout value (ms).',
		format  : 'nat',
		default : 20000
	},
	requestRetries : {
		doc     : 'Socket retries when a timeout occurs.',
		format  : 'nat',
		default : 3
	},
	// Mediasoup settings
	mediasoup :
	{
		numWorkers : {
			doc     : 'The number of Mediasoup workers to spawn. Defaults to the available CPUs count.',
			format  : 'nat',
			default : Object.keys(cpus()).length
		},
		worker :
		{
			logLevel : {
				doc     : 'The Mediasoup log level.',
				format  : String,
				default : 'warn'
			},
			logTags : {
				doc     : 'The Mediasoup log tags.',
				format  : Array,
				default : [
					'info',
					'ice',
					'dtls',
					'rtp',
					'srtp',
					'rtcp'
				]
			},
			rtcMinPort : {
				doc     : 'The Mediasoup start listening port number.',
				format  : 'port',
				default : 40000
			},
			rtcMaxPort : {
				doc     : 'The Mediasoup end listening port number.',
				format  : 'port',
				default : 49999
			}
		},
		// mediasoup Router settings.
		router :
		{
			// Router media codecs.
			mediaCodecs : {
				doc     : 'The Mediasoup codecs settings. [supportedRtpCapabilities](https://github.com/versatica/mediasoup/blob/v3/src/supportedRtpCapabilities.ts)',
				format  : '*',
				default :
				[
					{
						kind      : 'audio',
						mimeType  : 'audio/opus',
						clockRate : 48000,
						channels  : 2
					},
					{
						kind       : 'video',
						mimeType   : 'video/VP8',
						clockRate  : 90000,
						parameters :
						{
							'x-google-start-bitrate' : 1000
						}
					},
					{
						kind       : 'video',
						mimeType   : 'video/VP9',
						clockRate  : 90000,
						parameters :
						{
							'profile-id'             : 2,
							'x-google-start-bitrate' : 1000
						}
					},
					{
						kind       : 'video',
						mimeType   : 'video/h264',
						clockRate  : 90000,
						parameters :
						{
							'packetization-mode'      : 1,
							'profile-level-id'        : '4d0032',
							'level-asymmetry-allowed' : 1,
							'x-google-start-bitrate'  : 1000
						}
					},
					{
						kind       : 'video',
						mimeType   : 'video/h264',
						clockRate  : 90000,
						parameters :
						{
							'packetization-mode'      : 1,
							'profile-level-id'        : '42e01f',
							'level-asymmetry-allowed' : 1,
							'x-google-start-bitrate'  : 1000
						}
					}
				]
			}
		},
		// mediasoup WebRtcTransport settings.
		webRtcTransport :
		{
			listenIps : {
				doc     : 'The Mediasoup listen IPs. [TransportListenIp](https://mediasoup.org/documentation/v3/mediasoup/api/#TransportListenIp)',
				format  : Array,
				default : [
					{ ip: '0.0.0.0', announcedIp: null }
				]
			},
			initialAvailableOutgoingBitrate : {
				doc     : 'The Mediasoup initial available outgoing bitrate (in bps). [WebRtcTransportOptions](https://mediasoup.org/documentation/v3/mediasoup/api/#WebRtcTransportOptions)',
				format  : 'nat',
				default : 1000000
			},
			maxIncomingBitrate : {
				doc     : 'The Mediasoup maximum incoming bitrate for each transport. (in bps). [setMaxIncomingBitrate](https://mediasoup.org/documentation/v3/mediasoup/api/#transport-setMaxIncomingBitrate)',
				format  : 'nat',
				default : 1500000
			}
		}
	},
	// Prometheus exporter
	prometheus : {
		enabled : {
			doc     : 'Enables the Prometheus metrics exporter.',
			format  : 'Boolean',
			default : false
		},
		listen : {
			doc     : 'Prometheus metrics exporter listening address.',
			format  : 'String',
			default : 'localhost'
		},
		port : {
			doc     : 'The Prometheus metrics exporter listening port.',
			format  : 'port',
			default : 8889
		},
		// default metrics options
		deidentify : {
			doc     : 'De-identify IP addresses in Prometheus logs.',
			format  : 'Boolean',
			default : false
		},
		numeric : {
			doc     : 'Show numeric IP addresses in Prometheus logs.',
			format  : 'Boolean',
			default : false
		},
		quiet : {
			doc     : 'Include fewer labels in Prometheus metrics.',
			format  : 'Boolean',
			default : false
		},
		// aggregated metrics options
		period : {
			doc     : 'The Prometheus metrics exporter update period (seconds).',
			format  : 'nat',
			default : 15
		},
		secret : {
			doc     : 'The Prometheus metrics exporter authorization header: `Bearer <secret>` required to allow scraping.',
			format  : String,
			default : ''
		}
	},
	// User roles
	// All users have the role "NORMAL" by default. Other roles need to be
	// added in the "userMapping" function. The following accesses and
	// permissions are arrays of roles. Roles can be changed in userRoles.js
	//
	// Example:
	// [ userRoles.MODERATOR, userRoles.AUTHENTICATED ]
	accessFromRoles : {
		doc     : 'User roles.',
		format  : '*',
		default : {
			// The role(s) will gain access to the room
			// even if it is locked (!)
			[BYPASS_ROOM_LOCK] : [ userRoles.ADMIN ],
			// The role(s) will gain access to the room without
			// going into the lobby. If you want to restrict access to your
			// server to only directly allow authenticated users, you could
			// add the userRoles.AUTHENTICATED to the user in the userMapping
			// function, and change to BYPASS_LOBBY : [ userRoles.AUTHENTICATED ]
			[BYPASS_LOBBY]     : [ userRoles.NORMAL ]
		}
	},
	permissionsFromRoles : {
		doc     : 'User permissions from roles.',
		format  : '*',
		default : {
			// The role(s) have permission to lock/unlock a room
			[CHANGE_ROOM_LOCK] : [ userRoles.MODERATOR ],
			// The role(s) have permission to promote a peer from the lobby
			[PROMOTE_PEER]     : [ userRoles.NORMAL ],
			// The role(s) have permission to give/remove other peers roles
			[MODIFY_ROLE]      : [ userRoles.NORMAL ],
			// The role(s) have permission to send chat messages
			[SEND_CHAT]        : [ userRoles.NORMAL ],
			// The role(s) have permission to moderate chat
			[MODERATE_CHAT]    : [ userRoles.MODERATOR ],
			// The role(s) have permission to share audio
			[SHARE_AUDIO]      : [ userRoles.NORMAL ],
			// The role(s) have permission to share video
			[SHARE_VIDEO]      : [ userRoles.NORMAL ],
			// The role(s) have permission to share screen
			[SHARE_SCREEN]     : [ userRoles.NORMAL ],
			// The role(s) have permission to produce extra video
			[EXTRA_VIDEO]      : [ userRoles.NORMAL ],
			// The role(s) have permission to share files
			[SHARE_FILE]       : [ userRoles.NORMAL ],
			// The role(s) have permission to moderate files
			[MODERATE_FILES]   : [ userRoles.MODERATOR ],
			// The role(s) have permission to moderate room (e.g. kick user)
			[MODERATE_ROOM]    : [ userRoles.MODERATOR ]
		}
	},
	// Array of permissions. If no peer with the permission in question
	// is in the room, all peers are permitted to do the action. The peers
	// that are allowed because of this rule will not be able to do this 
	// action as soon as a peer with the permission joins. In this example
	// everyone will be able to lock/unlock room until a MODERATOR joins.
	allowWhenRoleMissing : {
		doc     : 'Allow when role missing.',
		format  : Array,
		default : [ CHANGE_ROOM_LOCK ]
	}
});

/**
 * Formats the schema documentation, calling the same function recursively.
 * @param docs the documentation object to extend
 * @param property the root property
 * @param schema the config schema fragment
 * @returns the documentation object
 */
function formatDocs(docs: any, property: string | null, schema: any)
{
	if (schema._cvtProperties)
	{
		Object.entries(schema._cvtProperties).forEach(([ name, value ]) =>
		{
			formatDocs(docs, `${property ? `${property}.` : ''}${name}`, value);
		});

		return docs;
	}

	if (property)
	{
		docs[property] = // eslint-disable-line no-param-reassign
		{
			doc     : schema.doc,
			format  : JSON.stringify(schema.format, null, 2),
			default : JSON.stringify(schema.default, null, 2)
		};
	}

	return docs;
}

// format docs
const configDocs = formatDocs({}, null, configSchema.getSchema());

let config: any = {};
let configError = '';
let configLoaded = false;

// Load config from file
for (const format of [ 'json', 'json5', 'yaml', 'yml', 'toml' ]) // eslint-disable-line no-restricted-syntax
{
	const filepath = `./config/config.${format}`;

	if (fs.existsSync(filepath))
	{
		try
		{
			logger.debug(`Loading config from ${filepath}`);
			configSchema.loadFile(filepath);
			configLoaded = true;
			break;
		}
		catch (err)
		{
			logger.debug(`Loading config from ${filepath} failed: ${err.message}`);
		}
	}
}

if (!configLoaded)
{
	logger.warn('No config file found, using defaults.');
	configSchema.load({});
}

// Perform validation
try
{
	configSchema.validate({ allowed: 'strict' });
	config = configSchema.getProperties();
}
catch (error: any)
{
	configError = error.message;
}

// load additional config module (no validation is performed)
if (fs.existsSync(`${__dirname}/../config/config.js`))
{
	try
	{
		const configModule = require('../config/config.js'); // eslint-disable-line @typescript-eslint/no-var-requires

		Object.assign(config, configModule);
	}
	catch (err)
	{
		logger.error(`Error loading ${config.configFile} module: ${err.message}`);
	}
}

// eslint-disable-next-line
// logger.debug('Using config:', config);

//
export {
	configSchema,
	config,
	configError,
	configDocs
};
